### ISSUE TITLE
Add a nice wrapper class to avoid implementing pipe-operator for custom views.

### ISSUE THREAD

**Issue by rrahn** [seqan/seqan3#142]
Please note, this is a design question/feature proposal.
Inspired by the nice tutorial about implementing your own ranges in the wiki, I was thinking that it would be great to take the burden off of the developers to always implement the pipe-operator for their custom views.
Especially, as it could be very confusing in the usage if stuff like `std::bind` together with placeholders is used to implement proper pipe-operator delegation for the respective view.
Hence, I came up with a CRTP pattern, that in general would allow us to implement a very handsome 
`my_view_fn` class where I don't need to think about the implementation of the pipe-operator and these kind of stuff.
Of course some stuff has to be dealt with, like proper const-fication, but in general I think that would be the way to go.
It might be possible, that I wasn't thinking of some corner cases so this should be discussed here, and we need to see if we can update the proposal or if we have to withdraw completely.

**[-note-begin-**
Of course if the view would have two constructors, while expecting 2 arguments and one of the constructors has a default parameter for the second argument, than there would be probably an ambiguous function call. Then again, the overloaded function call operator of `view_fn` could handle the default argument and thus making this design still applicable I think.
**-end-]**

But now enough talking. Lean back and enjoy! 😁 

```cpp

#include <vector>
#include <utility>
#include <memory>
#include <tuple>

namespace seqan3
{

// The factory that delegates the pipe to a custom call using the lhs of a |-operator
// implemented as a CRTP class.
template <typename derived_view_fn_t>
struct view_factory
{
    // A reference to the derived view_fn object.
    derived_view_fn_t & view_fn;

    // Construction.
    view_factory(derived_view_fn_t & derived) : view_fn(derived)
    {}

    // A special functor that gives us a named type for which we can overload the
    // pipe-operator. NOTE: I tried to use a lambda instead, but obviously, what should
    // the signature of the pipe operator look like? It would get two template parameters,
    // making it applicable for everything.
    template <typename... arg_ts>
    struct view_factory_fn
    {
        // Get access to the underlying view_fn.
        derived_view_fn_t & view_fn;

        // Store the variadic values somewhere.
        std::tuple<arg_ts...> member;

        // constructor.
        view_factory_fn(derived_view_fn_t & fn, arg_ts && ...args) :
            view_fn(fn),
            member{std::forward<arg_ts>(args)...}
        {}

        // helper class to unpack the tuple.
        template <typename rng_t, size_t... Is>
        auto explode(rng_t && rng, std::index_sequence<Is...> const &)
        {
            return view_fn(std::forward<rng_t>(rng), std::forward<arg_ts>(std::get<Is>(member))...);
        }

        // the operator called by the pipe-operator.
        auto operator()(auto && lhs_rng)
        {
            return explode(std::forward<decltype(lhs_rng)>(lhs_rng), std::make_index_sequence<sizeof...(arg_ts)>{});
        };
    };

    // the operator called by the derived class to get a proxy for calling the pipe-operator later on
    // without using the weird interface with std::bind.
    template <typename... arg_ts>
    auto operator()(arg_ts && ...args)
    {
        return view_factory_fn<arg_ts...>{view_fn, std::forward<arg_ts>(args)...};
    }

    // the actual magic pipe-operator with getting view_factory_fn on the rhs,
    // so that it does not get to be called for every type.
    template <typename rng_t, typename... arg_ts>
    friend auto operator|(rng_t && lhs, view_factory_fn<arg_ts...> && callable)
    {
        return callable(std::forward<rng_t>(lhs));
    }
};

// Some test view implementation that does not do a lot.
template <typename rng_t = void>
class my_view_impl
{
protected:

    std::shared_ptr<std::decay_t<rng_t>> data{};
    unsigned member1{};
public:
    my_view_impl() = default;

    my_view_impl(unsigned const m1) noexcept : member1(m1)
    {}

    my_view_impl(rng_t && rng, unsigned const m1) noexcept : data{std::make_shared<std::decay_t<rng_t>>(rng)}, member1{m1}
    {}

    unsigned get() const noexcept
    {
        return member1;
    }
};

// That would be an example .._fn implementation using the CRTP pattern,
// to get almost automatically access to the pipe notation.
struct my_view_fn : protected view_factory<my_view_fn>
{
    // the base class.
    using factory_t = view_factory<my_view_fn>;

    // the constructor.
    my_view_fn() : factory_t(*this)
    {};

    // Just tell the compiler there is a suitable operator() overload that apparently
    // does all the magic for us.
    using factory_t::operator();

    // Now specify what should happen if the final operator is called.
    // This allows us to use the functional call design or the pipe notation.
    template <typename rng_t>
    auto operator()(rng_t && rng, unsigned const p1)
    {
        return my_view_impl<rng_t>(std::forward<rng_t>(rng), p1);
    }
};

namespace view
{
// Well the code above is yet to be const-ified.
my_view_fn /*const*/ my_view;
}
}

using namespace seqan3;

// Some general use cases, that should work.
int main()
{
    { // use case 1: functional;
        std::vector<int> vec{{1, 2, 3, 4, 5}};
        auto v = view::my_view(vec, 4u);
        std::cout << v.get()  << "\n"; // should print 4
    }

    {  // use case 2: pipe version.
        std::vector<int> vec{{1, 2, 3, 4, 5}};
        auto v2 = vec | view::my_view(4u) | view::my_view(3u);
        std::cout << v.get()  << "\n"; // should print 3
    }
}
```

**Comment by h-2** [comment]
@sarahet says no!

**Comment by rrahn** [comment]
Like joke-no or broke-no!
What would be the rationale?
My thinking is purely guided by [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

**Comment by h-2** [comment]
Damn it, I had a long reply to this, but it somehow disappeared. I think it would be great to reduce complexity in writing view, but I am not sure the current solution does that.

The short version:

Abstractions are a double-bladed sword, they
:+1: hide uneccessary detail, re-use code, focus on functional changes 
:-1: make self-contained and self-explanatory code, suddenly depend on other code, defined somewhere else that is usually less intuitive that your original code.

If we want to introduce such an abstraction, the :+1: need to outweigh the :-1: . I was just cleaning up the nucleotides and I realised I could save some lines of code and repititions, if I had all nucleotides inherit from a CRTP-based facade. But it would definetely make a something very basic that new devs look at rather complex. And it wouldn't save a lot of code so I decided against it for now.

Let's look at your example:
```cpp
struct my_view_fn : protected view_factory<my_view_fn>
{
    // the base class.
    using factory_t = view_factory<my_view_fn>;

    // the constructor.
    my_view_fn() : factory_t(*this)
    {};

    // Just tell the compiler there is a suitable operator() overload that apparently
    // does all the magic for us.
    using factory_t::operator();

    // Now specify what should happen if the final operator is called.
    // This allows us to use the functional call design or the pipe notation.
    template <typename rng_t>
    auto operator()(rng_t && rng, unsigned const p1)
    {
        return my_view_impl<rng_t>(std::forward<rng_t>(rng), p1);
    }
};
```

VS

```cpp
struct my_view_fn
{
    // function-like interface: auto v = view::my(input, 3);
    template <typename irng_t>
    auto operator()(irng_t && irange, uint64_t const the_number) const
    {
        return view_add_number{std::forward<irng_t>(irange), the_number};
    }

    // helper for pipe notation, creates intermediate when calling view::my(3)
    auto operator()(uint64_t const the_number) const
    {
        return std::bind(my_view_fn(), std::placeholders::_1, the_number);
    }

    // access through pipe notation: auto v = input | view::my(3);
    // result of the intermediate is taken as right hand side argument
    template <typename irng_t>
    friend auto operator|(irng_t && irange, decltype(std::bind(my_view_fn(), std::placeholders::_1, 0ul)) const & bound_view)
    {
        return bound_view(std::forward<irng_t>(irange));
    }

};
```

it doesn't even save lines of code. One could argue that the itself is simpler, but that is a matter of debate. We now have four entities vs three. We don't have an std::bind, but we do have a friend declaration, CRTP,  operator inheritance... these are all not self-explanatory and what is more:
they now hide away the actual implementation in another file that you also need to understand if you want to understand what's going on.
Arguably the other file is even more complex than what we are replacing (which is not surprising, because abstractions need to be more general than concrete solutions).

For people who just want to use it and not understand it, it may not make a huge difference, they will copy and paste either way. Still I would argue that there is little benefit for them, because they don't even save lines that they are copying so it's "repeat yourself" but on a different level.

**Comment by h-2** [comment]
That is not saying I am against this in general, but I think we need more advantages before we introduce an abstraction that is this complex. Let's have a look at it together when we find the time!

**Comment by rrahn** [comment]
> it doesn't even save lines of code. One could argue that the itself is simpler, but that is a matter of debate. We now have four entities vs three. We don't have an std::bind, but we do have a friend declaration, CRTP, operator inheritance... these are all not self-explanatory and what is more:
they now hide away the actual implementation in another file that you also need to understand if you want to understand what's going on.

We could implement it in the facade using std::bind. That is just fine with me. But encapsulation is an essential entity of programming in like any language. It is what you need to do!!!!!
C++ give us just a lot more different ways to do it, depending on what your design goal is.
With regard to this I more and more cherish the design of SeqAn2 in many ways. It just lacked a proper documentation so people did do a lot of redundant things, they should not have done (including me). Most of this we can now circumvent by using actually real good C++ design patterns given the nice language features that we can use now. 
This means, that we need to accustom to language features that are out there quite a while but we never used. Hence, I misused the friend declaration in the prototype, and actually we don't need it here. 
Having said that, this could also be just a virtual inheritance, but this is not our desing goal.
Moreover, I think over time we will get more and more of these patterns especially rergarding views. I mean the ranges library is designed upon these principles and it totally makes sense.
If newbies don't grasp that, than either they are not willing to learn or there is a lack of good documentation, which makes it infinitely harder to learn the structure. I think the last point is the main thing we are going to prevent.

So why would you be reluctant about newbies learning proper C++ design patterns?
Right now, all they need to do is implement the constructor and make the function call operator() visible, and ... oh wait ... that's it.
And this is part of the design pattern!!! This is something to accustom too. Even if we never did it before. I think this is pretty standard in many other c++ libraries and not black magic.
And then they just have to implement one single function call operator for the current view implementation an redundant stuff is plugged in by the abstraction. 

>For people who just want to use it and not understand it, it may not make a huge difference, they will copy and paste either way. Still I would argue that there is little benefit for them, because they don't even save lines that they are copying so it's "repeat yourself" but on a different level.

Respectfully, but I think this argument is wrong in so many ways. 
Copy'n'paste is just wrong. Code duplication is just wrong. But if people copy the encapsulated version, they only have one place to update the code and by that reduce the potential errors that could happen by falsly copy'n'pasting.

I now start working more and more with views and the rationale behind that. Many things are much clearer now and a lot more will follow. So in the end i am certain we will have an encapsulation some how. I totally agree, that there might be possibly an even better interface. And maybe we figure it out. But as more and more views will follow, these structural changes would imply a lot of files to touch in order get them into the new structure. 
This is what DRY is for: it prevents these kind of issues in an early stage and keeps your software project agile and adaptable.

So to summarise, I really do not agree on your view of what you think abstraction takes away from you. That is IMHO just a wrong understanding of abstraction at all. I mean it is the fundamental strategy behind OOP. It is what we actually agreed upon to achieve, have we not. Only that we still do static over runtime polymorphism and that gives us certain patterns to use.

Ok sorry I probably went rouge some how. The whole thread/discussion drifted away i think.
Let me reformulate the issue here. We need this abstraction. I have made a proposal. This proposal can be discussed and adapted, not the fact that we need the abstraction.

And to calm ourselves, here is a 🍭.

**Comment by rrahn** [comment]
Maybe we can even use the already existing `pipe` infrastructure of the range library here.
I'll check if this is applicable for us.

**Comment by h-2** [comment]
Ok, just to clear this up: abstraction are good. Almost always. My only point was that using the abstraction should be significantly simpler and/or more compact (these two things can be the same, but don't have to be).

I have coded an abstraction with some inspiration from above, that I hope fulfils this:
```cpp
template <template <typename> class view_type>
struct define_view_functor
{
    template <typename irng_t, typename arg1_t, typename ...arg_types>
//         requires requires { view_type{irng_t{}, arg1_t{}, arg_types{}...}; }
    auto operator()(irng_t && irange, arg1_t && arg1, arg_types && ...args) const
    {
        return view_type<std::decay_t<irng_t>>{std::forward<irng_t>(irange),
                                               std::forward<arg1_t>(arg1),
                                               std::forward<arg_types>(args)...};
    }

    // binds the above iterator with placeholder for range (std::bind doesn't work here because of template template)
    template <typename ...arg_types>
    struct auxiliary_functor_t
    {
        // Store the arguments
        std::tuple<arg_types...> arguments;

        // helper function to unpack the tuple.
        template <typename irng_t, size_t... Is>
        auto explode(irng_t && irange, std::index_sequence<Is...> const &) const
        {
            using parent_type = define_view_functor<view_type, arg_types...>;
            return parent_type{}(std::forward<irng_t>(irange), static_cast<arg_types>(std::get<Is>(arguments))...);
            //TODO: need to investigate if std::get does the right thing
        }

        // the operator called by the pipe-operator.
        template <typename irng_t>
//             requires requires { view_type{irng_t{}, arg_types{}...}; }
        auto operator()(irng_t && irange) const
        {
            return explode(std::forward<irng_t>(irange), std::make_index_sequence<sizeof...(arg_types)>{});
        }
    };

    template <typename ...arg_types>
    auxiliary_functor_t<arg_types...> operator()(arg_types && ...args) const
    {
        return auxiliary_functor_t<arg_types...>{std::forward<arg_types>(args)...};
    }

    template <typename irng_t, typename ...arg_types>
        requires requires { view_type{irng_t{}, arg_types{}...}; }
    friend auto operator|(irng_t && irange, auxiliary_functor_t<arg_types...> const & bound_functor)
    {
        return bound_functor(std::forward<irng_t>(irange));
    }
};

// this the only thing the developer has to do: [not even parameter types need to be specified]
using add_number_fn = define_view_functor<view_add_number>;
```

I will also add a version for when you have no arguments.

Caveats:
* We cannot do real concept checks, neither on the range type, nor on the argument types (in cases where they are also template arguments; because we cannot pass a concept as a template parameter. a class template as parameter works, but a variable template (or concept) is simply not supported. However we can do a constructability test of the actual view and this should contain all the concept checks. I am doing this above.
* The actual view class must be able to deduce all it's template parameters from it's constructor arguments (automatically or via deduction guides)

**Comment by rrahn** [comment]
That's the spirit I am talking about.
